File: functions/src/domain/entities/conversation.ts
import {ConversationGroup} from "./conversation-group";


export class Conversation {

    constructor(
       public readonly conversationId: string,
       public readonly participants: string[],
       public readonly group?: ConversationGroup,
    ) {}

    get isGroup () : boolean {
        return this.group != null;
    }

    removedParticipants(conversationBeforeUpdate:Conversation) : Array<string> {
        return conversationBeforeUpdate.participants
            .filter((uid) => conversationBeforeUpdate.participants.includes(uid))
            .filter((uid) => !this.participants.includes(uid));
    }
}

File: flutter_app/lib/features/chat/domain/entities/conversation.dart

class Conversation {
  late final String conversationId;

  /// ID's of the members of the conversation
  final List<String> participants;

  /// User IDS of the users who are typing at the same time
  List<String> typingUids;

  bool get isGroup => group != null;

  final ConversationGroup? group;

  Conversation({
    required this.conversationId,
    required this.participants,
    required this.typingUids,
    this.group,
  });

}

class ConversationGroup {
  String title;
  final List<String> adminUids;
  final String createdBy;
  Map<String,DateTime> joinedAt;

  ConversationGroup({required this.title, required this.joinedAt, required this.adminUids, required this.createdBy});
}